### PR TITLE
event recorder

### PR DESCRIPTION
**Pull request by njuettner**
Introduce event recorder, which allows to operators to create events.

Usually event recorder could implement a logger, however I don't think this makes sense because our own logger is really special. It would only write the event to the logs additionally.

Towards: https://github.com/giantswarm/giantswarm/issues/11368

## Checklist

- [x] Update changelog in CHANGELOG.md.
- [ ] Update roadmap in ROADMAP.md.

### CODE REVIEW

**Comment by njuettner** [comment]
The general idea is to use the EventRecorder in each operator resource where it make sense, e.g. aws-operator resources.

In order to do that we need create a new EventRecorder for each controller (component). 

When we create a new controller from operatorkit, all of the config is not exposed so it can only used internally. 
My idea is to to call recorder.New in operators before calling `newXResources()` so we can inject the recorder for each resource.

File: controller/controller.go [hunk 1]
```diff
@@ -241,8 +256,28 @@ func (c *Controller) Reconcile(req reconcile.Request) (reconcile.Result, error)
...
+	}
+
+	res, err := c.reconcile(ctx, req, obj)
 	if err != nil {
+		// In case of microerror create an event on the object
+		if merr, ok := microerror.Cause(err).(*microerror.Error); ok {
+			if merr.Kind != "" && merr.Desc != "" {
+				c.recorder.Event(obj, corev1.EventTypeWarning, merr.Kind, merr.Desc)
+			}
+		}
```
Comment (xh3b4sd) [inline controller/controller.go:280]: Bonus points for making it pretty. 

```suggestion
		c.event.Emit(ctx, obj, err)
```

File: controller/internal/recorder/recorder.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+// New creates an event recorder to send custom events to Kubernetes to be recorded for targeted Kubernetes objects
+func New(c Config) record.EventRecorder {
+	eventBroadcaster := record.NewBroadcaster()
+	if _, isfake := c.K8sClient.(*k8sclienttest.Clients); !isfake {
+		eventBroadcaster.StartRecordingToSink(
+			&typedcorev1.EventSinkImpl{
+				Interface: c.K8sClient.K8sClient().CoreV1().Events("")})
+	}
+	return eventBroadcaster.NewRecorder(c.K8sClient.Scheme(), clientv1.EventSource{Component: c.Component})
```
Comment (xh3b4sd) [inline controller/internal/recorder/recorder.go:24]: How did the logging issue resolve now? Or is it still a thing?

Comment (njuettner) [reply controller/internal/recorder/recorder.go]: Yeah sorry for confusion, good news no logging needed. I don't know why my test wasn't working 🤷 .

File: controller/controller.go (hunk 1 above)
Comment (njuettner) [reply controller/controller.go]: I like that 👍

File: controller/controller.go [hunk 3]
```diff
@@ -159,6 +161,16 @@ func New(config Config) (*Controller, error) {
 		}
 	}
 
+	var eventRecorder recorder.Interface
+	{
+		c := recorder.Config{
+			Component: config.Name,
+			K8sClient: config.K8sClient,
```
Comment (xh3b4sd) [inline controller/controller.go:168]: `Component` is a setting and `K8sClient` is a dependency. See https://github.com/giantswarm/fmt/blob/master/go/struct_initialization.md. 

```suggestion
			K8sClient: config.K8sClient,

			Component: config.Name,
```

File: controller/controller.go [hunk 4]
```diff
@@ -241,8 +254,25 @@ func (c *Controller) Reconcile(req reconcile.Request) (reconcile.Result, error)
...
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{}, microerror.Mask(err)
+	}
+
+	res, err := c.reconcile(ctx, req, obj)
 	if err != nil {
+		// Microerror creates an error event on the object when kind and description is set.
+		c.event.Emit(ctx, obj, err)
+
```
Comment (xh3b4sd) [inline controller/controller.go:275]:
```suggestion
```

File: controller/internal/recorder/event.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	corev1 "k8s.io/api/core/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+type Recorder struct {
```
Comment (xh3b4sd) [inline controller/internal/recorder/event.go:12]: Some opinions. 

* Let's rename `service.go` to `recorder.go`. That service naming is not good as we learned in the past.
* Let's move `Recorder` to `recorder.go`. The struct and constructor magic should be in the same file. 
* Let's move the `Emit` method to `recorder.go`. No reason to keep 7 lines in a separate file.

File: controller/internal/recorder/service.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+	K8sClient k8sclient.Interface
+}
+
+// New creates an event recorder to send custom events to Kubernetes to be recorded for targeted Kubernetes objects.
+func New(c Config) Interface {
+	eventBroadcaster := record.NewBroadcaster()
+	if _, isfake := c.K8sClient.(*k8sclienttest.Clients); !isfake {
+		eventBroadcaster.StartRecordingToSink(
+			&typedcorev1.EventSinkImpl{
+				Interface: c.K8sClient.K8sClient().CoreV1().Events("")})
```
Comment (xh3b4sd) [inline controller/internal/recorder/service.go:22]:
```suggestion
				Interface: c.K8sClient.K8sClient().CoreV1().Events(""),
			},
		)
```

File: controller/internal/recorder/service.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+type Config struct {
+	Component string
+	K8sClient k8sclient.Interface
+}
+
+// New creates an event recorder to send custom events to Kubernetes to be recorded for targeted Kubernetes objects.
+func New(c Config) Interface {
+	eventBroadcaster := record.NewBroadcaster()
+	if _, isfake := c.K8sClient.(*k8sclienttest.Clients); !isfake {
```
Comment (xh3b4sd) [inline controller/internal/recorder/service.go:19]:
```suggestion

	_, isfake := c.K8sClient.(*k8sclienttest.Clients)
	if !isfake {
```

File: controller/internal/recorder/recorder.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+// New creates an event recorder to send custom events to Kubernetes to be recorded for targeted Kubernetes objects.
+func New(c Config) Interface {
+	eventBroadcaster := record.NewBroadcaster()
+	_, isfake := c.K8sClient.(*k8sclienttest.Clients)
+	if !isfake {
+		eventBroadcaster.StartRecordingToSink(
+			&typedcorev1.EventSinkImpl{
+				Interface: c.K8sClient.K8sClient().CoreV1().Events(""),
```
Comment (kopiczko) [inline controller/internal/recorder/recorder.go:32]: Fake doesn't support it?

File: controller/internal/recorder/recorder.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+			},
+		)
+	}
+	return &Recorder{
+		eventBroadcaster.NewRecorder(c.K8sClient.Scheme(), clientv1.EventSource{Component: c.Component}),
+	}
+}
+
+func (r *Recorder) Emit(ctx context.Context, obj pkgruntime.Object, err error) {
+	if merr, ok := microerror.Cause(err).(*microerror.Error); ok {
```
Comment (kopiczko) [inline controller/internal/recorder/recorder.go:42]: I'd prefer to use `errors.As` here. To be more prepared for the future.

File: integration/test/event/event_test.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	// wait a bit to let events appear on the cm
+	time.Sleep(5 * time.Second)
+
+	events, err := wrapper.Events(cm.GetNamespace())
+	if err != nil {
+		t.Fatal("failed to get events from configmap", nil, "got", err)
+	}
+	if len(events) == 0 {
```
Comment (kopiczko) [inline integration/test/event/event_test.go:121]: Wouldn't that be more strict?

```suggestion
	if len(events) != 2 {
```

File: controller/internal/recorder/recorder.go (hunk 8 above)
Comment (njuettner) [reply controller/internal/recorder/recorder.go:32]: No, not that I'm aware of.

File: integration/test/event/event_test.go (hunk 10 above)
Comment (njuettner) [reply integration/test/event/event_test.go]: Thanks, changed it but it's != 1 due to the fact the same event happened and it doesn't add a new line.

File: controller/internal/recorder/recorder.go (hunk 9 above)
Comment (njuettner) [reply controller/internal/recorder/recorder.go]: Make sense 👍

**Pull request merged**